kratos/python/add_tensor_adaptors_to_python.cpp

```diff
@@ -0,0 +1,276 @@
+    std::vector<std::size_t> strides(c_shape.size());
+
+    std::size_t stride_items = 1;
+    for (int i = c_shape.size() - 1; i >= 0; --i) {
+        strides[i] = sizeof(primitive_data_type) * stride_items;
+        stride_items *= c_shape[i];
+    }
+
+    // do nothing in the release of the numpy array since the ownership is not passed
+    // the ownership is kept with the TensorAdaptor.
+    pybind11::capsule release(rTensorAdaptor.ViewData().data(), [](void* a) {});
+
+    return numpy_array_type(
+        c_shape,
+        strides,
+        rTensorAdaptor.ViewData().data(),
+        release
+    );
+}
+
+template<class TTensorAdaptorType, class... TArgs>
+void SetNumpyArray(
+    TTensorAdaptorType& rTensorAdaptor,
+    std::variant<pybind11::array_t<TArgs, pybind11::array::c_style> const *...> pArray)
+{
+    std::visit([&rTensorAdaptor](const auto pArray) {
+        auto& r_array = *pArray;
+
+        KRATOS_ERROR_IF(r_array.ndim() == 0)
+            << "Passed data is not compatible [ array = "
+            << r_array << ", Tensor adaptor = " << rTensorAdaptor << " ].\n";
+
+        std::vector<int> shape(r_array.ndim());
+        std::copy(r_array.shape(), r_array.shape() + r_array.ndim(), shape.begin());
+
+        const auto& r_shape = rTensorAdaptor.Shape();
+
+        KRATOS_ERROR_IF_NOT(shape.size() == r_shape.size())
+            << "Dimensions mismatch. [ Tensor dimensions = " << r_shape.size()
+            << ", numpy array dimensions = " << shape.size()
+            << ", Tensor adaptor = " << rTensorAdaptor << " ].\n";
+
+        for (unsigned int i = 0; i < shape.size(); ++i) {
+            KRATOS_ERROR_IF_NOT(r_shape[i] == shape[i])
+                << "Shape mismatch. [ Tensor shape = " << rTensorAdaptor.Shape()
+                << ", numpy array shape = " << shape
+                << ", Tensor adaptor = " << rTensorAdaptor << " ].\n";
+        }
+
+        // copy data from the input to the Adaptor
+        IndexPartition<IndexType>(rTensorAdaptor.ViewData().size()).for_each([&r_array, &rTensorAdaptor](const auto Index) {
+            rTensorAdaptor.ViewData()[Index] = static_cast<typename TTensorAdaptorType::PrimitiveDataType>(r_array.data()[Index]);
+        });
+    }, pArray);
+}
+
+template<class TTensorAdaptorType>
+void AddBaseTensorAdaptor(
+    pybind11::module& rModule,
+    const std::string& AdaptorName)
+{
+    using container_type = typename TTensorAdaptorType::ContainerType;
+    using primitive_data_type = typename TTensorAdaptorType::PrimitiveDataType;
+    pybind11::class_<TTensorAdaptorType, typename TTensorAdaptorType::Pointer, Detail::TensorAdaptorTrampoline<container_type, primitive_data_type>>(rModule, AdaptorName.c_str())
+        .def("CollectData", &TTensorAdaptorType::CollectData)
+        .def("StoreData", &TTensorAdaptorType::StoreData)
+        .def("GetContainer", &TTensorAdaptorType::GetContainer)
+        .def("Shape", &TTensorAdaptorType::Shape)
```

Comment:
do you need the "Shape" method in python? you could just do ViewData.Shape ...

(indeed not a big problem if you do want to expose it)